### PR TITLE
[FSDP2] Used multi-grad hook when no inputs require grad

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132610
* __->__ #129259
* #131949

This PR uses `register_multi_post_accumuluate_grad_hook` to run the post-backward logic when the module inputs do not require gradient.

For context, FSDP2 currently relies on a hook on the module inputs that require gradient to run the post-backward logic (like a module full backward hook except with pytree support). This means that if none of the module inputs require gradient, then the post-backward logic is deferred to the final callback that runs at the end of backward, which may not be timely in some hybrid parallelism cases (e.g. a sparse arch before the FSDP dense arch). To address this case, we use this new multi-post-accumulate-grad hook.

Since whether the module inputs requiring grad could be dynamic from iteration to iteration, we guard this on a flag to make sure that we use the existing logic if the module inputs do require gradient.

<details>
<summary> Common Case </summary>

In the common case, the only module whose forward inputs do not require gradient is the root FSDP module, where the root FSDP module is the overall model's root. In this case, the multi-grad hook simply moves the root's post-backward from the final callback to a preceding `AccumulateGrad` node.

New:
Screenshot 2024-06-21 at 3 17 06 PM

Old:
Screenshot 2024-06-21 at 3 16 45 PM

</details>

cc @XilunWu @H-Huang @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @penguinwu @tianyu-l @yf225 @chauhang

Differential Revision: [D59012616](https://our.internmc.facebook.com/intern/diff/D59012616)

### CODE REVIEW

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 1]
```diff
@@ -556,3 +584,35 @@ def forward(ctx, param_group: FSDPParamGroup, *inputs: torch.Tensor):
     def backward(ctx, *grads: torch.Tensor):
         ctx.param_group.post_backward()
         return (None,) + grads
+
+
+def _register_multi_post_acc_grad_hook(
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:589]: @soulitzer 😓

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 2]
```diff
@@ -556,3 +584,35 @@ def forward(ctx, param_group: FSDPParamGroup, *inputs: torch.Tensor):
     def backward(ctx, *grads: torch.Tensor):
         ctx.param_group.post_backward()
         return (None,) + grads
+
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:587]: For context, the existing multi-grad hook does not work so well with the existing `post_backward` because the last parameter's `.grad` is not assigned yet when the mulit-grad hook runs (expected behavior).

We actually prefer this mulit-post-accumulate-grad hook, where the last parameter's `.grad` has already been assigned.

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 3]
```diff
@@ -444,7 +459,20 @@ def _register_post_backward_hook(
...
+            # Only need to register once since the unsharded parameter objects
+            # are preserved from iteration to iteration in eager
+            if self._multi_grad_hook_handle is None:
+                tensors = [
+                    fsdp_param.unsharded_param for fsdp_param in self.fsdp_params
+                ]
+                self._multi_grad_hook_handle = _register_multi_post_acc_grad_hook(
+                    tensors, self._multi_grad_post_backward
+                )
             return args, kwargs  # no tensors that require gradients
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:474]: The key idea is that before, we just return when there are no module inputs that require grad, falling back to the final callback that runs _at the end of backward_ to run post-backward for this module.

This multi-post-acc-grad hook allows running the module's post-backward earlier (but still correctly).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D59012616).

**Comment by awgu** [comment]
To-do: consider adding a test where the input requires grad changes from iteration to iteration

**Review by weifengpy — APPROVED** [review]

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py [hunk 4]
```diff
@@ -494,7 +509,25 @@ def _register_post_backward_hook(
...
+            # any unsharded parameters are frozen (where we conservatively do
+            # not run this hook in case the trainability is changing)
+            self._run_multi_grad_hook = (
+                all(t.requires_grad for t in unsharded_params)
+                and len(unsharded_params) > 0
+            )
+            # Only need to register once since the unsharded parameter objects
+            # are preserved from iteration to iteration in eager
+            if self._run_multi_grad_hook and self._multi_grad_hook_handle is None:
+                self._multi_grad_hook_handle = register_multi_post_accumulate_grad_hook(
```
Comment (awgu) [inline torch/distributed/_composable/fsdp/_fsdp_param_group.py:526]: to-do: Dynamo is erroring on this 🤔 
cc: @yf225

File: torch/distributed/_composable/fsdp/_fsdp_param_group.py (hunk 3 above)
Comment (sanketpurandare) [reply torch/distributed/_composable/fsdp/_fsdp_param_group.py:529]: This is neat!

**Review by sanketpurandare — APPROVED** [review]
This is very clean! Kudos!
